File: hIndex_bucketsort.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n=len(citations)
        bucket=[0]*(n+1)
        #bucket sort
        for i in range(n):
            if(citations[i]>=n): #
                bucket[n]+=1
            else:
                bucket[citations[i]]+=1
        sum1=0
        for i in range(n,-1,-1):
            sum1+=bucket[i]
            if (sum1>=i):
                return i
        

File: H-Index/H-Index_BucketSort.java
/* Time Complexity : O(n)) 
 *  n - len of array - citations */
/* Space Complexity : O(n) */
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this :


class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        //Bucket sort
        int[] buckets = new int[n+1];
        for(int i = 0; i < n; i++){
            if(citations[i] >= n){
                buckets[n]++;
            }else{
                buckets[citations[i]]++;
            }
        }

        int sum = 0;
        for(int i = n; i >= 0; i--){
            sum += buckets[i];
            //If sum is equal to idx or if there is a flip/switch ie., sum is greater than idx - we have our h-index
            if(sum >= i){
                return i;
            }
        }
        return 0;
    }
}